training/src/main/java/wolox/training/controllers/BookController.java

```diff
@@ -1,21 +1,66 @@
 package wolox.training.controllers;
 
-
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Controller;
 import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.DeleteMapping;
 import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestParam;
-
+import org.springframework.web.bind.annotation.ResponseStatus;
+import wolox.training.exceptions.BookIdMismatchException;
+import wolox.training.exceptions.BookNotFoundException;
+import wolox.training.models.Book;
+import wolox.training.repositories.BookRepository;
 
 @Controller
 public class BookController {
 
+    @Autowired
+    BookRepository bookRepository;
+
     @GetMapping("/greeting")
     public String greeting(@RequestParam(name="name", required=false, defaultValue="World") String name, Model model){
 
         model.addAttribute("name", name);
         return "greeting";
     }
 
+    @GetMapping("/")
+    public Iterable findAll(){
+        return bookRepository.findAll();
+    }
+
+    @GetMapping("/{id}")
+    public Book findOne(@PathVariable Long id) throws BookNotFoundException {
+        return bookRepository.findById(id).orElseThrow(BookNotFoundException::new);
+    }
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Book create(@RequestBody Book book) {
+        return bookRepository.save(book);
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) throws BookNotFoundException {
+        bookRepository.findById(id).orElseThrow(BookNotFoundException::new);
+        bookRepository.deleteById(id);
+    }
+
+    @PutMapping("/{id}")
+    public Book updateBook(@RequestBody Book book, @PathVariable Long id) throws BookNotFoundException, BookIdMismatchException {
```

Comment:
Just a suggestion, what about making different methods for each book attribute (e.g. udpateAuthor, updateTitle, updateYear, etc.)? 🙂 That way you only need to pass one parameter that will be updated in the book that already exists instead of an entire book.

Reply:
Ok :) Should I leave this method in case I want to update many attributes of one book?

Reply:
One way to do it would be to make one method that receives all attributes as parameters and make them optional, with a default value. That way you can update as many as you want without having to make separate methods 😄

Reply:
The idea of this is that now you could create and endpoint for each attribute. But that won't scale. The ideal thing is to have just one endpoint what updates the whole thing.
I would leave that as you have, and updating every attribute inside of that method.
The only cases where a sepparate update method is needed, is for example with a User's password, as that usually get's updated via a different use case.

Reply:
Also because otherwise the frontend would need to do N requests when updating N attributes of the entity.